oeps/oep-0016.rst

```diff
@@ -0,0 +1,261 @@
+allows pages to be converted one at a time.
+
+An investigation was performed to see whether Bootstrap components could be used
+on non-Bootstrap pages, but the conclusion was that the conflicts made this more
+trouble than it was worth. The recommendation is to convert pages in one shot
+which will be quicker and will have fewer issues.
+
+Reference Implementation
+========================
+
+In order to prove the viability of Bootstrap 4, a proof-of-concept was built
+that converted three different pages from the LMS:
+
+* The course bookmarks page which was a simple first page to target
+* The main courseware page which contains a number of custom styling rules
+* The new course outline page which is a typical `edX Pattern Library`_ page
+
+The following work was done:
+
+* A new lms-main-v3.scss file was introduced that used Bootstrap
+* An edX Bootstrap theme was implemented and included as a partial
+* The three pages were switched over to use the v3 styling
+* A "uses_bootstrap" context parameter enables a template to add Bootstrap JS
+* The header was reimagined with Bootstrap components
+* The page styles were updated to use Bootstrap mixins and theme variables
+* RTL support was provided using `postcss`_ and the `rtlcss`_ plugin.
+
+For the courseware, a new lms-course-v3.scss was added
+
+* This file included all of the course-specific Sass partials
+* All global-level rules were excluded to allow Bootstrap's styles to be used
+
+See the `Bootstrap Proof-of-Concept PR`_ for more details, including screenshots
+of the three converted pages. Note: the proof-of-concept was mostly completed
+during a two-day hackathon, thus demonstrating how quickly these conversions can
+be done.
+
+Rejected Alternatives
+=====================
+
+edX Pattern Library
+-------------------
+
+The `edX Pattern Library`_ is an initiative that was started in 2015 to
+modernize the way that edX applications were built. At the time, Bootstrap was
+considered to not be suitable for edX because applications always looked like
+they were built with Bootstrap. In addition, Bootstrap's rules were defined
+using Less and so were not compatible with edX's Sass-based styles. Finally, all
+sizing was done using pixels which was not compatible with responsive
+typography.
+
+The pattern library project was somewhat successful, and a number of edX
+pages were built using it. However, it became clear that there were not enough
+resources available to build a comprehensive solution, and the non-standard
+nature of the project meant that the community chose not to adopt it. It was
+at this point that the `edX Front End Working Group`_ decided to investigate
+other alternatives.
+
+For more information, see the wiki page `State of the Pattern Library`_.
+
+Material Design
+---------------
+
+Google's Material Design is another very successful web framework that was
+evaluated. It was ultimately considered to be too opinionated to support the
```

Comment:
I don't remember when this conclusion was drawn. And I personally think Material design was conceived for a [wide variety of use cases](https://material.io/guidelines/) for material design, specifically

* [Accessibility](https://material.io/guidelines/usability/accessibility.html)
* [Bidirectionality](https://material.io/guidelines/usability/bidirectionality.html)

[Material Components](https://github.com/material-components/material-components-web) are also kept up to date by Google.